### PR TITLE
Initial support for C++/CLI 

### PR DESCRIPTION
**Pull request by wli3**

### CODE REVIEW

**Comment by nguerrera** [comment]
FYI, @olgaark this is where we'll be making progress

**Comment by wli3** [comment]
@nguerrera get chance to review this?

**Comment by nguerrera** [comment]
Starting to look now

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.BeforeCommon.targets
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.BeforeCommon.targets]: This won't be necessary for C++ if we're disabling nuget.

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.BeforeCommon.targets
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.BeforeCommon.targets]: This won't be necessary for C++ if we're disabling nuget.

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.BeforeCommon.targets
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.BeforeCommon.targets]: This also isn't necessary for C++ .net core case. I am wondering if we should reduce this down to the things that we really need for C++. We should discuss with @dsplaisted

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets]: Shouldn't this check if Version is already set?

File: src/Assets/TestProjects/NetCoreCsharpAppReferenceCppCliLib/NETCoreCppCliTest/Directory.Build.props
Comment (nguerrera) [inline src/Assets/TestProjects/NetCoreCsharpAppReferenceCppCliLib/NETCoreCppCliTest/Directory.Build.props]: Are any of these no longer needed as of latest 16.4 build?

File: src/Tasks/Common/Resources/xlf/Strings.it.xlf
Comment (nguerrera) [inline src/Tasks/Common/Resources/xlf/Strings.it.xlf]: Bug here in xliff-tasks. I've seen it a few times now and really need to get to the bottom of it. Looks like a race condition. The translations got reverted.

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets]: Do the others need this too? _GetComHostPaths?

**Comment by wli3** [comment]
We also need to think about if the name of `UsingMicrosoftNETSdk` is good enough

File: src/Assets/TestProjects/NetCoreCsharpAppReferenceCppCliLib/NETCoreCppCliTest/Directory.Build.props
Comment (wli3) [reply src/Assets/TestProjects/NetCoreCsharpAppReferenceCppCliLib/NETCoreCppCliTest/Directory.Build.props]: Good catch and I think there is mismatch between names.

VC++ expecting Sdks\Microsoft.NET.Sdk\targets\core\Microsoft.NET.Sdk.Core.BeforeCommon.targets

we have Sdks\Microsoft.NET.Sdk\sdk\Sdk.BeforeCommon.targets

I will send email

File: src/Tasks/Common/Resources/xlf/Strings.it.xlf
Comment (wli3) [reply src/Tasks/Common/Resources/xlf/Strings.it.xlf]: i will revert this file and run generation again

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.BeforeCommon.targets
Comment (wli3) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.BeforeCommon.targets]: Removed this

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.BeforeCommon.targets
Comment (wli3) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.BeforeCommon.targets]: pending a meeting with Daniel

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets
Comment (wli3) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets]: added condition on Version being empty

File: src/Assets/TestProjects/NetCoreCsharpAppReferenceCppCliLib/NETCoreCppCliTest/Directory.Build.props
Comment (wli3) [reply src/Assets/TestProjects/NetCoreCsharpAppReferenceCppCliLib/NETCoreCppCliTest/Directory.Build.props]: per Olga's email. We cannot remove it yet(blocked on this chance check in and build, and Olga do the change on their side)

File: src/Tasks/Common/Resources/xlf/Strings.it.xlf
Comment (wli3) [reply src/Tasks/Common/Resources/xlf/Strings.it.xlf]: done

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets
Comment (wli3) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets]: added explicit dependency to Comhost and apphost

**Comment by wli3** [comment]
After all done i will squash most of the commits

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.BeforeCommon.targets
Comment (nguerrera) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.BeforeCommon.targets]: What I mean here is that this is doing special work for targeting net framework, but that is not something we will be supporting for the C++ feature. We should check what happens now if you put TargetFramework=net46 in vcxproj. I think it will break as the features that are importing SDK in VC project targets are doing so along with netcore specific compiler switches, etc.?

Comment (wli3) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.BeforeCommon.targets]: remove it for now. and have better error message when there is fullframework tfm

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.BeforeCommon.targets
Comment (wli3) [inline src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.BeforeCommon.targets]: UsingNETSdkDefaults

Comment (wli3) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.BeforeCommon.targets]: "classic" -> "C++/CLI which don't use SDK style defaults"

Comment (wli3) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.BeforeCommon.targets]: All "classic" is changed

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.BeforeCommon.targets
Comment (wli3) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.BeforeCommon.targets]: Removed. And added check for fullframework

File: src/Tasks/Common/Resources/Strings.resx
Comment (wli3) [inline src/Tasks/Common/Resources/Strings.resx]: I am not sure the error is good enough. The direct reason is UsingNETSdkDefaults=false + using full framework. However, from the user point of view. CLRSupport =netcore + targetframework=net4XX is the problem

File: src/Tasks/Microsoft.NET.Build.Tasks/sdk/Sdk.props
Comment (wli3) [inline src/Tasks/Microsoft.NET.Build.Tasks/sdk/Sdk.props]: Added UsingNETSdkDefaults=true here. Comments looks good?

**Comment by wli3** [comment]
@nguerrera I addressed comments, check again?

File: src/Tasks/Common/Resources/Strings.resx
Comment (nguerrera) [reply src/Tasks/Common/Resources/Strings.resx]: I would remove the "below 5".

**Comment by nguerrera** [comment]
> After all done i will squash most of the commits

We should reword 8cdfd15 as this is not really what we have at this point, it is specifically being done for C++.

**Review by nguerrera — APPROVED** [review]

File: src/Tasks/Common/Resources/Strings.resx
Comment (nguerrera) [reply src/Tasks/Common/Resources/Strings.resx]: .NET 5 is not .NET Framework 5

**Comment by wli3** [comment]
I'll just change it to "Support targeting .NET Core from C++/CLI which don't use Microsoft.NET.Sdk"

**Comment by wli3** [comment]
cleaned commit message, will merge once CI is done

File: src/Tasks/Common/Resources/Strings.resx
Comment (wli3) [reply src/Tasks/Common/Resources/Strings.resx]: Removed "below 5"

**Pull request merged**
